File: src/lib/oop/Object/MemberList.r
#ifndef MEMBERLIST_H
#define MEMBERLIST_H

#include "Legacy_MemberList.r"

// Forward declaration of incomplete type
typedef struct memberList MemberList;

struct memberList {

    /// `MemberList` implements `Legacy_ObjectComponent`.
    Legacy_ObjectComponent *legacyObjectComponent;

    /// `MemberList` implements `AutoDestructable`.
    AutoDestructable *autoDestructable;

    Legacy_MemberList *legacyMemberList;

    Legacy_Object *(*addMemberWhichIsLegacy_Object)(
            MemberList *self, char *memberName, Legacy_Object *legacyObject);
    Legacy_Object *(*addMemberWhichIsPrimitive)(
            MemberList *self, char *memberName,
            void *dynamicallyAllocatedPrimitive);
    Legacy_Object *(*getMemberByName)(MemberList *self, char *memberName);
};

MemberList *MemberListConstructor();

#endif //MEMBERLIST_H

File: src/lib/oop/Object/MemberList.c
#include "MemberList.r"

//// TODO: maybe remove. could be redundant.
//Legacy_Object *
//addMemberWhichIsPrimitiveAnonymousPointer(MemberList *self, char *memberName) {
//    return self->addMemberWhichIsPrimitive(
//            self, memberName, calloc(1, sizeof(TYPEOF_ANONYMOUS_POINTER)));
//}

Legacy_Object *getMemberByName_MemberList(MemberList *self, char *memberName) {
    return self->legacyMemberList->getMemberByName(self->legacyMemberList,
                                                   memberName);
}

void *MemberListDestructor(MemberList *memberList) {
    if (memberList == NULL) { return NULL; }

    AutoDestructableDestructor(memberList->autoDestructable);

    memberList->legacyMemberList->legacyObjectComponent->destructable
            ->destructor(memberList->legacyMemberList);

    free(memberList->legacyObjectComponent);
    free(memberList);

    return NULL;
}

MemberList *MemberListConstructor() {
    MemberList *instance = calloc(1, sizeof *instance);
    if (instance == NULL) { /* error handling here */
    }

    instance->addMemberWhichIsLegacy_Object =
            &addMemberWhichIsLegacy_ObjectContainer;
    instance->addMemberWhichIsPrimitive = &addMemberWhichIsPrimitive;
    instance->getMemberByName           = &getMemberByName_MemberList;

    instance->legacyObjectComponent =
            Legacy_ObjectComponentConstructorClassName("MemberList");

    static Constructable const constructable = {
            .constructor = (void *(*const)(void) )(&MemberListConstructor)};
    instance->legacyObjectComponent->constructable = &constructable;

    static Destructable const destructable = {
            .destructor = (void *(*const)(void *) )(&MemberListDestructor)};
    instance->legacyObjectComponent->destructable = &destructable;

    instance->legacyMemberList = Legacy_MemberListConstructor();

    return instance;
}